go-controller/pkg/ovnwebhook/nodeadmission.go

```diff
@@ -0,0 +1,169 @@
+		}
+		if v.action == changed {
+			return fmt.Errorf("%s cannot be changed once set", util.OvnNodeChassisID)
+		}
+		return nil
+	},
+	util.OvnNodeZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to "global" or <nodeName> initially
+		if v.action == added && (v.value == types.OvnDefaultZone || v.value == nodeName) {
+			return nil
+		}
+
+		// it is allowed for the annotation to be changed to <nodeName>
+		if v.action == changed && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s or %s, it cannot be removed", util.OvnNodeZoneName, types.OvnDefaultZone, nodeName)
+	},
+}
+
+// interconnectNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users in IC environments
+var interconnectNodeAnnotationChecks = map[string]checkNodeAnnot{
+	util.OvnNodeMigratedZoneName: func(v annotationChange, nodeName string) error {
+		// it is allowed for the annotation to be set to <nodeName>
+		if (v.action == added || v.action == changed) && v.value == nodeName {
+			return nil
+		}
+
+		return fmt.Errorf("%s can only be set to %s, it cannot be removed", util.OvnNodeMigratedZoneName, nodeName)
+	},
+}
+
+// hybridOverlayNodeAnnotationChecks holds annotations allowed for ovnkube-node:<nodeName> users hybrid overlay environments
+var hybridOverlayNodeAnnotationChecks = map[string]checkNodeAnnot{
+	hotypes.HybridOverlayDRMAC: nil,
+	hotypes.HybridOverlayDRIP:  nil,
+}
+
+type NodeAdmission struct {
+	annotationChecks map[string]checkNodeAnnot
+	annotationKeys   sets.Set[string]
+}
+
+func NewNodeAdmissionWebhook(enableInterconnect, enableHybridOverlay bool) *NodeAdmission {
+	checks := make(map[string]checkNodeAnnot)
+	maps.Copy(checks, commonNodeAnnotationChecks)
+	if enableInterconnect {
+		maps.Copy(checks, interconnectNodeAnnotationChecks)
+	}
+	if enableHybridOverlay {
+		maps.Copy(checks, hybridOverlayNodeAnnotationChecks)
+	}
+	return &NodeAdmission{
+		annotationChecks: checks,
+		annotationKeys:   sets.New[string](maps.Keys(checks)...),
+	}
+}
+
+var _ admission.CustomValidator = &NodeAdmission{}
+
+func (p NodeAdmission) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore deletion, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	nodeName, isOVNKubeNode := ovnkubeNodeIdentity(req.UserInfo)
+
+	if !isOVNKubeNode {
```

Comment:
this might be a stupid question... how does this admission work with accounts other than ovnkube-node? Like what if a cluster admin needs to modify some value because he downgraded/upgraded or some other reason. Will this webhook block them as well?

Reply:
In the current implementation the webhook only checks ovnkube-node users (`isOVNKubeNode == true`).
I can change the code to not allow other users to modify the annotations that are managed by ovnkube-node but please note that this is only protecting us from `node/status` updates, if someone uses `node` resource to update the annotation it will not go through our webhook.

Reply:
so i thought in our discussion today we said we wanted the same validation to apply to other credentials that reside on the node, like multus? Here we are allowing anyone else to do node/status updates, but we want to enforce this for any user trying to do a status update dont we?

Reply:
That was the plan, after talking to @jcaamano I realized that this is not so straightforward. In multi-homing, localnet and layer2 network controllers set the podnetworks annotation from cluster-manager. I will try to think how we can sort that out while still blocking other users from messing with our annotations.

Reply:
I added an extra parameter that allows us to specify extra users that can modify ovnkube-notes annotations.